gitclone/gitclone_test.go

```diff
@@ -0,0 +1,411 @@
+package gitclone
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"testing"
+
+	"github.com/bitrise-io/bitrise-init/step"
+	"github.com/bitrise-io/go-utils/command"
+	"github.com/bitrise-io/go-utils/command/git"
+)
+
+const always = 9999
+const rawCmdError = "dummy_cmd_error"
+
+func Test_checkoutState(t *testing.T) {
+	tests := []struct {
+		name       string
+		cfg        Config
+		cmdOutputs map[string]commandOutput
+		want       *step.Error
+		wantCmds   []string
+	}{
+		{
+			name: "Checkout commit",
+			cfg: Config{
+				RepositoryURL: "https://github.com/bitrise-io/steps-git-clone.git",
+				Commit:        "76a934ae80f12bb9b504bbc86f64a1d310e5db64",
+			},
+			want: nil,
+			wantCmds: []string{
+				`git "fetch"`,
+				`git "checkout" "76a934ae80f12bb9b504bbc86f64a1d310e5db64"`,
+			},
+		},
+		{
+			name: "Checkout commit, branch specified",
+			cfg: Config{
+				RepositoryURL: "https://github.com/bitrise-io/steps-git-clone.git",
+				Commit:        "76a934ae80f12bb9b504bbc86f64a1d310e5db64",
+				Branch:        "hcnarb",
+			},
+			want: nil,
+			wantCmds: []string{
+				`git "fetch"`,
+				`git "checkout" "76a934ae80f12bb9b504bbc86f64a1d310e5db64"`,
+			},
+		},
+		{
+			name: "Checkout commit with retry",
+			cfg: Config{
+				RepositoryURL: "https://github.com/bitrise-io/steps-git-clone.git",
+				Commit:        "76a934ae80f12bb9b504bbc86f64a1d310e5db64",
+			},
+			cmdOutputs: map[string]commandOutput{
+				`git "fetch"`: {failForCalls: 1},
+			},
+			want: nil,
+			wantCmds: []string{
+				`git "fetch"`,
+				`git "fetch"`,
+				`git "checkout" "76a934ae80f12bb9b504bbc86f64a1d310e5db64"`,
+			},
+		},
+		// Commit with unshallow
+		{
+			name: "Checkout branch",
+			cfg: Config{
+				RepositoryURL: "https://github.com/bitrise-io/steps-git-clone.git",
+				Branch:        "hcnarb",
+			},
+			want: nil,
+			wantCmds: []string{
+				`git "fetch" "origin" "refs/heads/hcnarb"`,
+				`git "checkout" "hcnarb"`,
+				`git "merge" "origin/hcnarb"`,
+			},
+		},
+		{
+			name: "Checkout nonexistent branch",
+			cfg: Config{
+				RepositoryURL: "https://github.com/bitrise-io/steps-git-clone.git",
+				Branch:        "fake",
+			},
+			cmdOutputs: map[string]commandOutput{
+				`git "fetch" "origin" "refs/heads/fake"`: {failForCalls: always},
```

Comment:
You could also use a [mocking package](https://github.com/stretchr/testify/#mock-package) to stub the mock runner to return various outputs for various inputs. Example for setting up this particular mocking package: https://github.com/bitrise-steplib/steps-generic-file-storage/blob/90872ec8e32248f5bbc4ce695e86c4db398ff269/bulkfiledownloader/mockFileDownloader_test.go#L25

Reply:
Do you agree to this?

Reply:
I'll create a new PR showcasing how that would look like.